authorization/client: implement owner cache for services

jkralik opened the pull request.

jkralik committed:
authorization/client: implement owner cache for services

ondrejtomcik commented on authorization/client/ownerCache.go:
```diff
@@ -0,0 +1,354 @@
+	subscription     *nats.Subscription
+	devices          SortedSlice
+	validUntil       time.Time
+	devicesWasSynced bool
```
plural

ondrejtomcik commented on authorization/client/ownerCache.go:
```diff
@@ -0,0 +1,354 @@
+	wg   sync.WaitGroup
+}
+
+// NewOwnerCache creates owner cache.
```
Well, that's obvious. But what is the owner's cache? If you add a comment, do it not only to pass the check but to give some more information that new creates a new instance.

Danielius1922 committed:
Move sortedSlice to pkg and create tests

Danielius1922 commented on authorization/client/ownerCache.go:
```diff
@@ -0,0 +1,355 @@
+	deviceIDs = strings.MakeSortedSlice(deviceIDs)
+	var j int
+	for i := range deviceIDs {
+		if j >= len(d.devices) || deviceIDs[i] != d.devices[j] {
```
is devicesIds guaranteed to contains everything in d.devices?

because for example:

deviceIDs = ["dev2", "dev3", "dev4"]
d.devices = ["dev1", "dev2", "dev3", "dev4"]

===============
added = ["dev2", "dev3", "dev4"]

which doesn't seem right?

Danielius1922 commented on authorization/client/ownerCache.go:
```diff
@@ -0,0 +1,355 @@
+	}
+	j = 0
+	for i := range d.devices {
+		if j >= len(deviceIDs) || d.devices[i] != d.devices[j] {
```
`d.devices[i] != d.devices[j]` ? always false unless i == j

Danielius1922 committed:
authorization: fix deviceIds processing

Update calculation of added and removed to deviceIds
to correct algorithm using set difference operation.

Danielius1922 committed:
ownerCache: minor code updates
